nodes/stretch_ik_control.py

```diff
@@ -0,0 +1,1217 @@
+        vel = None
+        while not check_cancel() and not self.__reached_termination(
+            termination, err, vel, cartesian_mask
+        ):
+            self.node.get_logger().debug(" Loop Started...")
+            # Leave early if ROS shuts down or timeout is reached
+            if not rclpy.ok() or remaining_time(
+                self.node.get_clock().now(), start_time, timeout
+            ) <= Duration(seconds=0.0):
+                yield MotionGeneratorRetval.FAILURE
+                return
+
+            # Get the current joint state
+            q = self.__get_q(joint_position_overrides, all_joints=True)
+            self.node.get_logger().debug(
+                f" Joint Positions: {list(zip(self.all_joints, q))}"
+            )
+
+            # Get the error between the end-effector pose and the goal pose
+            ok, err = self.get_err(
+                goal_odom,
+                remaining_time(self.node.get_clock().now(), start_time, timeout),
+                q_all=q,
+            )
+            if not ok:
+                yield MotionGeneratorRetval.FAILURE
+                return
+            if err_callback is not None:
+                err_callback(err)
+            self.node.get_logger().debug(f" Error: {err}")
+
+            # Get the Jacobian matrix for the chain
+            J = pinocchio.computeFrameJacobian(
+                self.ik_solver.model,
+                self.ik_solver.data,
+                q,
+                self.ik_solver.ee_frame_idx,
+                pinocchio.ReferenceFrame.LOCAL_WORLD_ALIGNED,
+            )
+            self.node.get_logger().debug(f" Jacobian: {J}")
+
+            # Mask the Jacobian matrix to only include the articulated joints
+            J[:, non_articulated_joints_mask] = 0.0
+            if cartesian_mask is not None:
+                J[np.logical_not(cartesian_mask), :] = 0.0
+
+            # Calculate the pseudo-inverse of the Jacobian
+            J_pinv = np.linalg.pinv(J, rcond=1e-6)
+            self.node.get_logger().debug(f" Jacobian Pseudo-Inverse: {J_pinv}")
+
+            # Re-mask the pseudo-inverse to address any numerical issues
+            J_pinv[non_articulated_joints_mask, :] = 0.0
+            if cartesian_mask is not None:
+                J_pinv[:, np.logical_not(cartesian_mask)] = 0.0
+
+            # Calculate the joint velocities
+            vel = self.K @ J_pinv @ err
+            self.node.get_logger().debug(
+                f" Joint Velocities: {list(zip(self.controllable_joints, vel))}"
+            )
+
+            # Execute the velocities
+            self.__execute_velocities(vel, move_base, move_arm, rate_hz)
+
+            # Yield control back to the main action loop
+            yield MotionGeneratorRetval.CONTINUE
+
+        yield MotionGeneratorRetval.FAILURE if check_cancel() else MotionGeneratorRetval.SUCCESS
+        return
+
+    def set_navigation_mode(self, timeout: Duration, rate_hz: float = 10.0) -> bool:
```

Comment:
this logic feels like it's going to appear in a lot of applications, would be nice to factor out

Reply:
I kept it in for now, but am happy to take lead on factoring it out when we know where we want to put it or have other applications to use it.